OSDOCS-13459#Add cross-subscription support for Azure File (#92108)
openshift/openshift-docs: modules/persistent-storage-csi-azure-file-cross-sub-dynamic-pre-provisioning-pv-pvc-procedure.adoc

// Module included in the following assemblies:
//
// * storage/container_storage_interface/persistent_storage-csi-azure-file.adoc
//
:_mod-docs-content-type: PROCEDURE
[id="persistent-storage-csi-azure-file-cross-sub-dynamic-pre-provisioning-pv-pvc-procedure_{context}"]
= Pre-provisioning across subscriptions for Azure File by creating a PV and PVC:

.Prerequisites

> * Installed Azure cluster with the service principal or managed identity as Azure identity

Reviewer: This seems to need something. 
```suggestion
* Installed Azure cluster with the service principal or managed identity as the /an Azure identity
```

* Access to another subscription in the tenant with the cluster  

* Logged in to the Azure CLI

.Procedure

> . For the desired existing Azure File share, record the resource group, storage account, storage account key, and Azure File name. These values are used for the next steps.

Reviewer: Per [ISG](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#d), do not use _desire_, use _want_. However, I don't like _wanted_ here. I see _your_ earlier in the preview. Can you come up with a better word/rewrite? 
```suggestion
. For your Azure File share, record the resource group, storage account, storage account key, and Azure File name. These values are used for the next steps.
```

. Create a secret for the persistent volume parameter `spec.csi.nodeStageSecretRef.name` by running the following command:
+

> [source, terminal]
> ----
> $ oc create secret generic azure-storage-account-{storageaccount-name}-secret --from-literal=azurestorageaccountname="<azure-storage-account-name>" --from-literal azurestorageaccountkey="<azure-storage-account-key>" --type=Opaque <1>
> ----
> <1> Where `<azure-storage-account-name>` and `<azure-storage-account-key>` are the Azure storage account name and key respectively that you recorded in Step 1.

Reviewer: Maybe [this format](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#callouts-in-code-blocks)?

```suggestion
[source, terminal]
----
$ oc create secret generic azure-storage-account-{storageaccount-name}-secret --from-literal=azurestorageaccountname="<azure-storage-account-name>" --from-literal azurestorageaccountkey="<azure-storage-account-key>" --type=Opaque
----
+
--
where:
 `<azure-storage-account-name>` and `<azure-storage-account-key>` are the Azure storage account name and key respectively that you recorded in Step 1.
--
```

. Create a persistent volume (PV) using a similar configuration to the following example file:
+

> .Example PV YAML file
> [source, terminal]

Reviewer: ```suggestion
.Example PV YAML file
[source,yaml]
```

kind: PersistentVolume
metadata:
  annotations:
    pv.kubernetes.io/provisioned-by: file.csi.azure.com
  name: <pv-name> <1>
spec:
  capacity:
    storage: 10Gi <2>
  accessModes:
    - ReadWriteMany
  persistentVolumeReclaimPolicy: Retain  
  storageClassName: azurefile-csi-manual
  mountOptions:
    - cache=strict
    - nosharesock
    - actimeo=30
    - nobrl 
  csi:
    driver: file.csi.azure.com
    volumeHandle: "{resource-group-name}#{storage-account-name}#{file-share-name}" <3>
    volumeAttributes:
      shareName: <existing-file-share-name> <4>
    nodeStageSecretRef:
      name: <secret-name>  <5>
      namespace: <secret-namespace>  <6>
----
<1> <pv-name> is the PV name.
<2> The size of the PV.
<3> Ensure that `volumeHandle` is unique for every identical share in the cluster.
<4> For `<existing-file-share-name>, use only the file share name and not the full path.
<5> <secret-name> is the secret name created in the previous step.
<6> <secret-namespace> is the namespace where the secret resides.

. Create a persistent value claim (PVC) specifying the existing Azure File share referenced in Step 1 using a similar configuration to the following:
+
.Example PVC YAML file

> [source, yaml]

Reviewer: Not sure the space is needed, or does anything.

```suggestion
[source,yaml]
```

// Module included in the following assemblies:
//
// * storage/container_storage_interface/persistent_storage-csi-azure-file.adoc
//
:_mod-docs-content-type: PROCEDURE
[id="persistent-storage-csi-azure-file-cross-sub-dynamic-pre-provisioning-pv-pvc-procedure_{context}"]
= Static provisioning across subscriptions for Azure File by creating a PV and PVC:

.Prerequisites
* Installed {product-title} cluster on Azure with the service principal or managed identity as an Azure identity in one subscription (call it Subscription A)

* Access to another subscription (call it Subscription B) with the storage that is in the same tenant as the cluster

* Logged in to the Azure CLI

.Procedure
. For your Azure File share, record the resource group, storage account, storage account key, and Azure File name. These values are used for the next steps.

. Create a secret for the persistent volume parameter `spec.csi.nodeStageSecretRef.name` by running the following command:
+
[source, terminal]
----

> $ oc create secret generic azure-storage-account-{storageaccount-name}-secret --from-literal=azurestorageaccountname="<azure-storage-account-name>" --from-literal azurestorageaccountkey="<azure-storage-account-key>" --type=Opaque

Reviewer 2: we have a mix of <> and {} in that command
we also have {storageaccount-name} and <azure-storage-account-name> which are the same thing IIUC
if so let's keep <azure-storage-account-name> for both

Author: I've changed to <> since that's what we generally use.

----
+
Where:
`<azure-storage-account-name>` and `<azure-storage-account-key>` are the Azure storage account name and key respectively that you recorded in Step 1.

. Create a persistent volume (PV) by using a similar configuration to the following example file:
+
.Example PV YAML file
[source,terminal]
----
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    pv.kubernetes.io/provisioned-by: file.csi.azure.com
  name: <pv-name> <1>
spec:
  capacity:
    storage: 10Gi <2>
  accessModes:
    - ReadWriteMany
  persistentVolumeReclaimPolicy: Retain  
  storageClassName: <sc-name> <3>
  mountOptions:
    - cache=strict
    - nosharesock
    - actimeo=30
    - nobrl 
  csi:
    driver: file.csi.azure.com
    volumeHandle: "{resource-group-name}#{storage-account-name}#{file-share-name}" <4>
    volumeAttributes:
      shareName: <existing-file-share-name> <5>
    nodeStageSecretRef:
      name: <secret-name>  <6>
      namespace: <secret-namespace>  <7>
----
<1> The name of the PV.
<2> The size of the PV.

> <3> The storage class name.

Reviewer 2: In case of static provisioning the SC name does not really matter it does not need to actually exists right? Should we call it out explicitly to avoid confusion ? i.e avoid customers wondering if they should create one.

@duanwei33

Reviewer 3: I think it matters, the sc name in PV and PVC helps binding, it could help avoid unexpected behaviour. (Like there are other azure-disk PVs without sc name and the size is the same)
But yes, the storage class itself (with the same name) exists or not will not impact the binding, I think for customer scenario,  non-admin users have no access to list PV, so usually an admin will create a sc which a meaningful name and create pvs with the same sc name inside, then the non-admin users create the PVC with the sc for expected pvs.

In short, let's keep sc name in both pvc and pv.

Reviewer 2: ok to keep it but it can be misleading as customers will not know what to set here.

Author: so.....leave as is or change?

Reviewer 4: I personally think we should keep the SC in the example + comment that it's used to disable dynamic provisioning for such PV / PVC.

> What happens if a user sets an existing (unrelated) storage class?

In theory nothing, the PVC will try to bind to an existing PV. In practice, when the manually created PV and PVC accidentally don't match, Kubernetes will try to dynamically provision a new PV. Which is not what the user wants. Either a dedicated SC with `provisioner: kubernetes.io/no-provisioner` or non-existing SC will ensure the dynamic provisioning does not happen. Mis-matched PV + PVC stay Pending and user can fix their errors.

Reviewer 2: Agree to keep the SC in the PVC/PV definition i was just wondering if we could add an extra note on the fact that the SC should not actually to avoid potential user errors (or as Jan said create a no-provisioner one). Not mentioning anything can be error prone i think

Reviewer 2: excellent, way more clear for users now, thanks

// Module included in the following assemblies:
//
// * storage/container_storage_interface/persistent_storage-csi-azure-file.adoc
//
:_mod-docs-content-type: PROCEDURE
[id="persistent-storage-csi-azure-file-cross-sub-dynamic-pre-provisioning-pv-pvc-procedure_{context}"]

> = Pre-provisioning across subscriptions for Azure File by creating a PV and PVC:

Reviewer 2: It's usually called static provisioning rather than pre-provisioning
@RomanBednar @duanwei33 thoughts?

Reviewer 3: In my understanding, these two words can be used interchangeably for the mentioned scenario. Sure, “static provisioning” is okay to me.

Author: I think in our content, we are generally using the term "static", so to be consistent, I'll change it to static.
\

.Prerequisites

> * Installed Azure cluster with the service principal or managed identity as an Azure identity

Author: Done.

> * Installed Azure cluster with the service principal or managed identity as an Azure identity

Reviewer 2: Do we mean OCP cluster on Azure? Is it standard policy to call it "Azure Cluster"

Author: I think you're right. I've changed it.

> * Installed Azure cluster with the service principal or managed identity as an Azure identity in one subscription (call it Subscription A)

Author: I think so. I've changed it.

> * Access to another subscription in the tenant with the cluster  

Author: Done.

> * Access to another subscription (call it Subscription B) in the same tenant with the cluster

Reviewer 2: same tenant "as" the cluster?

This subscription (B) should be the one with the storage IIUC, better call it out.

Author: Done.

  name: <pv-name> <1>
spec:
  capacity:
    storage: 10Gi <2>
  accessModes:
    - ReadWriteMany
  persistentVolumeReclaimPolicy: Retain  
  storageClassName: azurefile-csi-manual
  mountOptions:
    - cache=strict
    - nosharesock
    - actimeo=30
    - nobrl 
  csi:
    driver: file.csi.azure.com
    volumeHandle: "{resource-group-name}#{storage-account-name}#{file-share-name}" <3>
    volumeAttributes:
      shareName: <existing-file-share-name> <4>
    nodeStageSecretRef:
      name: <secret-name>  <5>
      namespace: <secret-namespace>  <6>
----
<1> The name of the PV.
<2> The size of the PV.
<3> Ensure that `volumeHandle` is unique for every identical share in the cluster.
<4> For `<existing-file-share-name>, use only the file share name and not the full path.
<5> The secret name created in the previous step.
<6> The namespace where the secret resides.

. Create a persistent value claim (PVC) specifying the existing Azure File share referenced in Step 1 using a similar configuration to the following:
+
.Example PVC YAML file
[source,yaml]
----
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: <pvc-name> <1>
spec:

>   storageClassName: <sc-name-cross-sub> <2>

Reviewer 3: Good catch, yes they should be the same!  And we can make this clear.
(My oversight in the draft doc)

>   storageClassName: <sc-name-cross-sub> <2>

Author: Fixed.
